### ISSUE TITLE
When close() is called on a link that got closed due to external factors, the listeners are not cleared

### ISSUE THREAD

**Issue by ramya-rao-a** [amqp/rhea-promise#30]
@amarzavery,

Say a link gets closed due to external factors (one way is to disable internet connection in the midst of receiving messages) i.e. no-one called `link.close()`, but `link.isOpen()` returns` false`.
What is the suggested way for consumers of `rhea-promise` to clean up the link?

[link.close()](https://github.com/amqp/rhea-promise/blob/1c2b7db133b667fb2b56f8e3cd36014c94aac0ce/lib/link.ts#L216) clears only the `close` and `error` event listeners that were attached inside the `close()` function, but not the listeners for all other events that were attached in the [_initializeEventListeners](https://github.com/amqp/rhea-promise/blob/1c2b7db133b667fb2b56f8e3cd36014c94aac0ce/lib/link.ts#L280) function.

Calling [link.remove()](https://github.com/amqp/rhea-promise/blob/1c2b7db133b667fb2b56f8e3cd36014c94aac0ce/lib/link.ts#L197) at this point does clear the listeners, but I am not sure if this is the right way to go about it.

I would expect `link.close()` to do the right thing here.

cc @ShivangiReja

**Comment by ShivangiReja** [comment]
Sample code:

```javascript
import {
  Connection, EventContext, ConnectionOptions, Receiver, ReceiverEvents
} from "rhea-promise";
import { ConnectionEvents } from "rhea";

const host = " ";
const username = "";
const password = "";
const port = 5671;
const receiverAddress = "";

async function main(): Promise<void> {
  const connectionOptions: ConnectionOptions = {
    transport: "tls",
    host: host,
    hostname: host,
    username: username,
    password: password,
    port: port,
    reconnect: false
  };

  const connection: Connection = new Connection(connectionOptions);
  await connection.open();

  const receiver: Receiver = await connection.createReceiver({
    name: "receiver-1",
    source: {
      address: receiverAddress
    },
  });

  receiver.on(ReceiverEvents.message, (context: EventContext) => {
    console.log("Received message: %O", context.message);
  });

  receiver.on(ReceiverEvents.receiverError, (context: EventContext) => {
    if (context.receiver && context.receiver.error) {
      console.log(">>>>> An error occurred for receiver: %O.", context.receiver.error);
    }
  });

  const disconnected: any = async (context: EventContext) => {
    console.log("Disconnected Event Fired");
    const connectionError = context.connection && context.connection.error
      ? context.connection.error
      : context.error;
    if (connectionError) {
      console.log("Error occurred on the amqp connection: %O", connectionError);
    }

    console.log(`Reciever is open?: ${receiver.isOpen()}`);
    console.log(`ListenerCount for message event: ${receiver.listenerCount(ReceiverEvents.message)}`);
    console.log(`ListenerCount for error event: ${receiver.listenerCount(ReceiverEvents.receiverError)}`);
    await receiver.close();
    console.log(`After close ListenerCount for message event: ${receiver.listenerCount(ReceiverEvents.message)}`);
    console.log(`After close ListenerCount for error event: ${receiver.listenerCount(ReceiverEvents.receiverError)}`);

  };

  connection.on(ConnectionEvents.disconnected, disconnected);
}

main().catch((err) => console.log(err));
```

**Comment by ramya-rao-a** [comment]
While using the above sample code to receive messages, disable the network connection. You will see the handler for the `disconnected` event being called and the console logs showing the below
- Receiver is not open
- There is 1 listener for the message and error event each
- These listeners remain even after `receiver.close()` is called.

**Comment by ramya-rao-a** [comment]
Related issue in rhea: https://github.com/amqp/rhea/issues/205

**Comment by ramya-rao-a** [comment]
As per discussions in https://github.com/amqp/rhea/issues/205, it is concluded that it is the consumer's responsibility to call `remove` on the links.

Keeping this issue open to track the need for `removeAllSessions` function on the connection object which will call the `remove_all_sessions` function on rhea's connection object thus allowing the consumer to clear things at one go.

cc @ShivangiReja

**Comment by ramya-rao-a** [comment]
this is fixed with #34 and #33
